File: up_tck_java/java_test_agent/src/main/java/org/tck/SocketRPCClient.java
package org.tck;

import org.eclipse.uprotocol.rpc.CallOptions;
import org.eclipse.uprotocol.rpc.RpcClient;
import org.eclipse.uprotocol.transport.builder.UAttributesBuilder;
import org.eclipse.uprotocol.v1.*;

import java.io.IOException;
import java.net.InetSocketAddress;
import java.net.Socket;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class SocketRPCClient {
    private Socket socket;

    public SocketRPCClient(String hostIp, int port, Socket socket) throws IOException {
        if (socket != null) {
            this.socket = socket;
        } else {
            this.socket = new Socket();
            this.socket.connect(new InetSocketAddress(hostIp, port));
        }
    }

    private UMessage sendToServiceSocket(UUri topic, UPayload payload, CallOptions callOptions) throws IOException {

        UUID requestId = UUID.getDefaultInstance();
        byte[] requestIdB = requestId.toString().getBytes();
        UAttributesBuilder builder = UAttributesBuilder.request(UPriority.UPRIORITY_CS4, topic, callOptions.timeout());
        UAttributes attributes = builder.build();

        UMessage umsg = UMessage.newBuilder()
                .setSource(topic)
                .setAttributes(attributes)
                .setPayload(payload)
                .build();
        byte[] umsg_bytes = umsg.toByteArray();

        socket.getOutputStream().write(umsg_bytes);

        while (true) {
            int msg_len = 32767;
            byte[] recv_data = new byte[msg_len];
            int readBytes = socket.getInputStream().read(recv_data);
            if (readBytes == -1) {
                continue;
            }
            UMessage responseUmsg = null;
            try {
                responseUmsg.parseFrom(recv_data);
            } catch (UnsupportedOperationException ue) {
                System.out.println("Data isn't type UMessage: \n" + ue.toString());
                continue;
            } catch (Exception err) {
                throw new RuntimeException(err);
            }
            if(responseUmsg.hasAttributes()) {
                UUID response_reqid = responseUmsg.getAttributes().getReqid();
                if (requestId.equals(response_reqid)) {
                    System.out.println("Got response: \n" + responseUmsg.toString());
                    return responseUmsg;
                } else {
                    System.out.println("Response wasn't meant for current client: \n" + responseUmsg.toString());
                    continue;
                }
            }
        }
    }

    public CompletionStage<UMessage> invokeMethod(UUri topic, UPayload payload, CallOptions callOptions) {
        ExecutorService executor = Executors.newFixedThreadPool(1);

        CompletableFuture<UMessage> response = CompletableFuture.supplyAsync(() -> {
            try {
                return this.sendToServiceSocket(topic, payload, callOptions);
            } catch (IOException e) {
                throw new RuntimeException(e);
            }
        }, executor);

        executor.shutdown(); // Initiates an orderly shutdown in which previously submitted tasks are executed, but no new tasks will be accepted.
        return response;
    }
}

File: up_tck/up_client_socket_python/socket_rpcclient.py
# -------------------------------------------------------------------------
#
# Copyright (c) 2024 General Motors GTO LLC
#
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
# SPDX-FileType: SOURCE
# SPDX-FileCopyrightText: 2024 General Motors GTO LLC
# SPDX-License-Identifier: Apache-2.0
#
# -------------------------------------------------------------------------
import sys
import socket
from concurrent.futures import Future, ThreadPoolExecutor

from logger.logger import logger
from uprotocol.proto.uattributes_pb2 import UAttributes
from uprotocol.proto.umessage_pb2 import UMessage
from uprotocol.proto.upayload_pb2 import UPayload
from uprotocol.proto.uri_pb2 import UUri
from uprotocol.proto.uuid_pb2 import UUID
from uprotocol.rpc.rpcclient import RpcClient

from up_tck.python_utils.constants import BYTES_MSG_LENGTH


class SocketRPCClient(RpcClient):
    """
    RpcClient is an interface used by code generators for uProtocol services defined in proto files such as the core
    uProtocol services found in <a href=https://github.com/eclipse-uprotocol/uprotocol-core-api>here</a>.<br> The
    interface provides a
    clean contract for all transports to implement to be able to support RPC on their platform.<br> Each platform MUST
    implement this interface.<br> For more details please refer to<br>
    <a href=https://github.com/eclipse-uprotocol/uprotocol-spec/blob/main/up-l2/README.adoc>[RpcClient
    Specifications]</a>
    """

    def __init__(self, server_ip: str, server_port: int, server_conn: socket.socket = None) -> None:
        '''
        @param server_ip: the ip address for the socket to connect to
        @param port: the port for the socket to connect to
        @param socket: cli_socket connected socket

        '''
        if server_conn is not None:
            self.socket = server_conn
        else:
            self.socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            self.socket.connect((server_ip, server_port))

    def __send_to_service_socket(self, topic: UUri, payload: UPayload, attributes: UAttributes):
        """
        Sends the using the provided UMessage data then waits until it gets a response message.
        @param topic: sending UUri protobuf.
        @param payload: sending UPayload protobuf.
        @param attributes: sending UAttributes protobuf.
        @return: a response message from the receiving client/service.
        """

        # Get UAttributes's request id
        request_id: UUID = attributes.id
        request_id_b: bytes = request_id.SerializeToString()

        # Protobuf message

        if topic is not None:
            attributes.source.CopyFrom(topic)

        umsg = UMessage(attributes=attributes, payload=payload)

        # Now serialized, then send
        umsg_serialized: bytes = umsg.SerializeToString()

        self.socket.sendall(umsg_serialized)

        while True:
            # Wait and receive data from server
            recv_data: bytes = self.socket.recv(BYTES_MSG_LENGTH)

            if recv_data == b'':
                continue

            response_umsg: UMessage = UMessage()

            try:
                # Unpack received data into UMessage
                response_umsg.ParseFromString(recv_data)
            except AttributeError as ae:
                logger.info("Data isn't type UMessage: \n" + str(ae))
                continue
            except Exception as err:
                raise Exception(err)

            response_reqid: UUID = response_umsg.attributes.reqid

            logger.info(f"Req-id {request_id_b} vs. Resp-id {response_reqid.SerializeToString()}")

            if request_id_b == response_reqid.SerializeToString():
                # Got response
                logger.info('Got response: \n' + str(response_umsg))
                return response_umsg
            else:
                # If response wasn't meant for current client
                logger.info("Response wasn't meant for current client: \n" + str(response_umsg))
                continue

    def invoke_method(self, topic: UUri, payload: UPayload, attributes: UAttributes) -> Future:
        """
        Support for RPC method invocation.<br><br>

        @param topic: topic of the method to be invoked (i.e. the name of the API we are calling).
        @param payload:The request message to be sent to the server.
        @param attributes: metadata for the method invocation (i.e. priority, timeout, etc.)
        @return: Returns the CompletableFuture with the result or exception.
        """
        with ThreadPoolExecutor(max_workers=1) as executor:
            # Mark each future with its groups
            response: Future = executor.submit(self.__send_to_service_socket, topic, payload, attributes)

            return response
